File: examples/typescript/binarysearchanswerExamples.ts
// 1. Koko Eating Bananas
function minEatingSpeed(piles: number[], h: number): number {
    let left = 1, right = Math.max(...piles);
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        const hours = piles.reduce((sum, pile) => sum + Math.ceil(pile / mid), 0);
        if (hours <= h) right = mid;
        else left = mid + 1;
    }
    return left;
}

// 2. Capacity To Ship Packages Within D Days
function shipWithinDays(weights: number[], days: number): number {
    let left = Math.max(...weights), right = weights.reduce((a, b) => a + b);
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        let daysNeeded = 1, currentWeight = 0;
        for (const weight of weights) {
            if (currentWeight + weight > mid) {
                daysNeeded++;
                currentWeight = weight;
            } else {
                currentWeight += weight;
            }
        }
        if (daysNeeded <= days) right = mid;
        else left = mid + 1;
    }
    return left;
}

// 3. Split Array Largest Sum
function splitArray(nums: number[], m: number): number {
    let left = Math.max(...nums), right = nums.reduce((a, b) => a + b);
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        let subarrays = 1, currentSum = 0;
        for (const num of nums) {
            if (currentSum + num > mid) {
                subarrays++;
                currentSum = num;
            } else {
                currentSum += num;
            }
        }
        if (subarrays <= m) right = mid;
        else left = mid + 1;
    }
    return left;
}

// 4. Minimum Number of Days to Make m Bouquets
function minDays(bloomDay: number[], m: number, k: number): number {
    if (m * k > bloomDay.length) return -1;
    let left = Math.min(...bloomDay), right = Math.max(...bloomDay);
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        let bouquets = 0, consecutive = 0;
        for (const bloom of bloomDay) {
            if (bloom <= mid) {
                consecutive++;
                if (consecutive === k) {
                    bouquets++;
                    consecutive = 0;
                }
            } else {
                consecutive = 0;
            }
        }
        if (bouquets >= m) right = mid;
        else left = mid + 1;
    }
    return left;
}

// 5. Smallest Divisor Given Threshold
function smallestDivisor(nums: number[], threshold: number): number {
    let left = 1, right = Math.max(...nums);
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        const sum = nums.reduce((acc, num) => acc + Math.ceil(num / mid), 0);
        if (sum <= threshold) right = mid;
        else left = mid + 1;
    }
    return left;
}

// 6. Magnetic Force Between Two Balls
function maxDistance(position: number[], m: number): number {
    position.sort((a, b) => a - b);
    let left = 1, right = position[position.length - 1] - position[0];
    while (left < right) {
        const mid = Math.floor((left + right + 1) / 2);
        let count = 1, lastPos = position[0];
        for (let i = 1; i < position.length; i++) {
            if (position[i] - lastPos >= mid) {
                count++;
                lastPos = position[i];
                if (count === m) break;
            }
        }
        if (count >= m) left = mid;
        else right = mid - 1;
    }
    return left;
}

// 7. Minimize Max Distance to Gas Station
function minmaxGasDist(stations: number[], k: number): number {
    let left = 0, right = 1e8;
    while (right - left > 1e-6) {
        const mid = (left + right) / 2;
        let needed = 0;
        for (let i = 1; i < stations.length; i++) {
            needed += Math.floor((stations[i] - stations[i - 1]) / mid);
        }
        if (needed <= k) right = mid;
        else left = mid;
    }
    return left;
}

// 8. Find K-th Smallest Pair Distance
function smallestDistancePair(nums: number[], k: number): number {
    nums.sort((a, b) => a - b);
    let left = 0, right = nums[nums.length - 1] - nums[0];
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        let count = 0, leftPtr = 0;
        for (let rightPtr = 1; rightPtr < nums.length; rightPtr++) {
            while (nums[rightPtr] - nums[leftPtr] > mid) leftPtr++;
            count += rightPtr - leftPtr;
        }
        if (count >= k) right = mid;
        else left = mid + 1;
    }
    return left;
}

// 9. Median of Two Sorted Arrays
function findMedianSortedArrays(nums1: number[], nums2: number[]): number {
    if (nums1.length > nums2.length) [nums1, nums2] = [nums2, nums1];
    
    const m = nums1.length, n = nums2.length;
    let left = 0, right = m;
    
    while (left <= right) {
        const partitionX = Math.floor((left + right) / 2);
        const partitionY = Math.floor((m + n + 1) / 2) - partitionX;
        
        const maxLeftX = partitionX === 0 ? -Infinity : nums1[partitionX - 1];
        const minRightX = partitionX === m ? Infinity : nums1[partitionX];
        
        const maxLeftY = partitionY === 0 ? -Infinity : nums2[partitionY - 1];
        const minRightY = partitionY === n ? Infinity : nums2[partitionY];
        
        if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
            if ((m + n) % 2 === 0) {
                return (Math.max(maxLeftX, maxLeftY) + Math.min(minRightX, minRightY)) / 2;
            } else {
                return Math.max(maxLeftX, maxLeftY);
            }
        } else if (maxLeftX > minRightY) {
            right = partitionX - 1;
        } else {
            left = partitionX + 1;
        }
    }
    return -1;
}

// 10. Aggressive Cows
function aggressiveCows(stalls: number[], cows: number): number {
    stalls.sort((a, b) => a - b);
    let left = 1, right = stalls[stalls.length - 1] - stalls[0];
    while (left < right) {
        const mid = Math.floor((left + right + 1) / 2);
        let count = 1, lastPos = stalls[0];
        for (let i = 1; i < stalls.length; i++) {
            if (stalls[i] - lastPos >= mid) {
                count++;
                lastPos = stalls[i];
                if (count === cows) break;
            }
        }
        if (count >= cows) left = mid;
        else right = mid - 1;
    }
    return left;
}

// Test cases
console.log("=== Binary Search on Answer Examples ===");

// Test 1: Koko Eating Bananas
console.log(`Min Eating Speed([3,6,7,11], 8): ${minEatingSpeed([3,6,7,11], 8)}`);

// Test 2: Ship Within Days
console.log(`Ship Within Days([1,2,3,4,5,6,7,8,9,10], 5): ${shipWithinDays([1,2,3,4,5,6,7,8,9,10], 5)}`);

// Test 3: Split Array Largest Sum
console.log(`Split Array([7,2,5,10,8], 2): ${splitArray([7,2,5,10,8], 2)}`);

// Test 4: Min Days for Bouquets
console.log(`Min Days([1,10,3,10,2], 3, 1): ${minDays([1,10,3,10,2], 3, 1)}`);

// Test 5: Smallest Divisor
console.log(`Smallest Divisor([1,2,5,9], 6): ${smallestDivisor([1,2,5,9], 6)}`);

// Test 6: Magnetic Force
console.log(`Max Distance([1,2,3,4,7], 3): ${maxDistance([1,2,3,4,7], 3)}`);

// Test 7: Min Max Gas Distance
console.log(`Min Max Gas Dist([1,2,3,4,5,6,7,8,9,10], 9): ${minmaxGasDist([1,2,3,4,5,6,7,8,9,10], 9)}`);

// Test 8: Smallest Distance Pair
console.log(`Smallest Distance Pair([1,3,1], 1): ${smallestDistancePair([1,3,1], 1)}`);

// Test 9: Median of Two Sorted Arrays
console.log(`Find Median([1,3], [2]): ${findMedianSortedArrays([1,3], [2])}`);

// Test 10: Aggressive Cows
console.log(`Aggressive Cows([1,2,4,8,9], 3): ${aggressiveCows([1,2,4,8,9], 3)}`);

File: examples/python/binary_search_answer_examples.py
# 1. Min Eating Speed
def min_eating_speed(piles, h):
    left, right = 1, max(piles)
    while left < right:
        mid = (left + right) // 2
        if sum((pile + mid - 1) // mid for pile in piles) <= h:
            right = mid
        else:
            left = mid + 1
    return left

# 2. Ship Within Days
def ship_within_days(weights, days):
    left, right = max(weights), sum(weights)
    while left < right:
        mid = (left + right) // 2
        days_needed, current_weight = 1, 0
        for weight in weights:
            if current_weight + weight > mid:
                days_needed += 1
                current_weight = weight
            else:
                current_weight += weight
        if days_needed <= days:
            right = mid
        else:
            left = mid + 1
    return left

# 3. Split Array
def split_array(nums, m):
    left, right = max(nums), sum(nums)
    while left < right:
        mid = (left + right) // 2
        subarrays, current_sum = 1, 0
        for num in nums:
            if current_sum + num > mid:
                subarrays += 1
                current_sum = num
            else:
                current_sum += num
        if subarrays <= m:
            right = mid
        else:
            left = mid + 1
    return left

# 4. Min Days
def min_days(bloom_day, m, k):
    if m * k > len(bloom_day):
        return -1
    left, right = min(bloom_day), max(bloom_day)
    while left < right:
        mid = (left + right) // 2
        bouquets = consecutive = 0
        for bloom in bloom_day:
            if bloom <= mid:
                consecutive += 1
                if consecutive == k:
                    bouquets += 1
                    consecutive = 0
            else:
                consecutive = 0
        if bouquets >= m:
            right = mid
        else:
            left = mid + 1
    return left

# 5. Smallest Divisor
def smallest_divisor(nums, threshold):
    left, right = 1, max(nums)
    while left < right:
        mid = (left + right) // 2
        if sum((num + mid - 1) // mid for num in nums) <= threshold:
            right = mid
        else:
            left = mid + 1
    return left

# 6. Max Distance
def max_distance(position, m):
    position.sort()
    left, right = 1, position[-1] - position[0]
    while left < right:
        mid = (left + right + 1) // 2
        count, last_pos = 1, position[0]
        for i in range(1, len(position)):
            if position[i] - last_pos >= mid:
                count += 1
                last_pos = position[i]
                if count == m:
                    break
        if count >= m:
            left = mid
        else:
            right = mid - 1
    return left

# 7. Minmax Gas Dist
def minmax_gas_dist(stations, k):
    left, right = 0, 1e8
    while right - left > 1e-6:
        mid = (left + right) / 2
        needed = sum(int((stations[i] - stations[i-1]) / mid) for i in range(1, len(stations)))
        if needed <= k:
            right = mid
        else:
            left = mid
    return left

# 8. Smallest Distance Pair
def smallest_distance_pair(nums, k):
    nums.sort()
    left, right = 0, nums[-1] - nums[0]
    while left < right:
        mid = (left + right) // 2
        count = left_ptr = 0
        for right_ptr in range(1, len(nums)):
            while nums[right_ptr] - nums[left_ptr] > mid:
                left_ptr += 1
            count += right_ptr - left_ptr
        if count >= k:
            right = mid
        else:
            left = mid + 1
    return left

# 9. Find Median Sorted Arrays
def find_median_sorted_arrays(nums1, nums2):
    if len(nums1) > len(nums2):
        nums1, nums2 = nums2, nums1
    
    m, n = len(nums1), len(nums2)
    left, right = 0, m
    
    while left <= right:
        partition_x = (left + right) // 2
        partition_y = (m + n + 1) // 2 - partition_x
        
        max_left_x = float('-inf') if partition_x == 0 else nums1[partition_x - 1]
        min_right_x = float('inf') if partition_x == m else nums1[partition_x]
        
        max_left_y = float('-inf') if partition_y == 0 else nums2[partition_y - 1]
        min_right_y = float('inf') if partition_y == n else nums2[partition_y]
        
        if max_left_x <= min_right_y and max_left_y <= min_right_x:
            if (m + n) % 2 == 0:
                return (max(max_left_x, max_left_y) + min(min_right_x, min_right_y)) / 2
            else:
                return max(max_left_x, max_left_y)
        elif max_left_x > min_right_y:
            right = partition_x - 1
        else:
            left = partition_x + 1
    return -1

# 10. Aggressive Cows
def aggressive_cows(stalls, cows):
    stalls.sort()
    left, right = 1, stalls[-1] - stalls[0]
    while left < right:
        mid = (left + right + 1) // 2
        count, last_pos = 1, stalls[0]
        for i in range(1, len(stalls)):
            if stalls[i] - last_pos >= mid:
                count += 1
                last_pos = stalls[i]
                if count == cows:
                    break
        if count >= cows:
            left = mid
        else:
            right = mid - 1
    return left

def run_examples():
    print("=== Binary Search on Answer Examples ===")
    
    # Test 1: Koko Eating Bananas
    print(f"Min Eating Speed([3,6,7,11], 8): {min_eating_speed([3,6,7,11], 8)}")
    
    # Test 2: Ship Within Days
    print(f"Ship Within Days([1,2,3,4,5,6,7,8,9,10], 5): {ship_within_days([1,2,3,4,5,6,7,8,9,10], 5)}")
    
    # Test 3: Split Array Largest Sum
    print(f"Split Array([7,2,5,10,8], 2): {split_array([7,2,5,10,8], 2)}")
    
    # Test 4: Min Days for Bouquets
    print(f"Min Days([1,10,3,10,2], 3, 1): {min_days([1,10,3,10,2], 3, 1)}")
    
    # Test 5: Smallest Divisor
    print(f"Smallest Divisor([1,2,5,9], 6): {smallest_divisor([1,2,5,9], 6)}")
    
    # Test 6: Magnetic Force
    print(f"Max Distance([1,2,3,4,7], 3): {max_distance([1,2,3,4,7], 3)}")
    
    # Test 7: Min Max Gas Distance
    print(f"Min Max Gas Dist([1,2,3,4,5,6,7,8,9,10], 9): {minmax_gas_dist([1,2,3,4,5,6,7,8,9,10], 9)}")
    
    # Test 8: Smallest Distance Pair
    print(f"Smallest Distance Pair([1,3,1], 1): {smallest_distance_pair([1,3,1], 1)}")
    
    # Test 9: Median of Two Sorted Arrays
    print(f"Find Median([1,3], [2]): {find_median_sorted_arrays([1,3], [2])}")
    
    # Test 10: Aggressive Cows
    print(f"Aggressive Cows([1,2,4,8,9], 3): {aggressive_cows([1,2,4,8,9], 3)}")

if __name__ == "__main__":
    run_examples()